src/neighborsearch/filtered.jl

```diff
@@ -0,0 +1,146 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    FilteredSearch(method, nmax=Inf; maxpercategory, maxpersector, order=Euclidean())
+
+A method for searching at most `nmax` neighbors using `method`. Extra
+constraints available: `maxpercategory` and `maxpersector`. The priority is
+given to the nearest neighbor using `order`.
+
+## Max per category
+
+It can be a `NamedTuple` or `Dict`, where the first element is the property name
+and the second defines the max neighbor per available categories of the property.
+In the example below, no more than 2 neighbors from the same hole are selected.
+
+`FilteredSearch(method, maxpercategory = (holeid = 2,))`
+[ASCII illustration]
+
+## Max per sector
+
+Only allow a max number of neighbors inside each sector. The sectors are the
+cartesian quadrants (2-D) or octants (3-D). If the neighborhood is an `Ellipsoid`,
+the sectors will match the rotated quadrants/octants.
+
+`FilteredSearch(method, maxpersector = 2)`
+[ASCII illustration]
+"""
+struct FilteredSearch{M<:NeighborSearchMethod} <: BoundedNeighborSearchMethod
+  method::M
+  nmax::Int
+  maxpercategory
+  maxpersector
+  order
+end
+
+FilteredSearch(method::M, nmax=0; maxpercategory=nothing,
+               maxpersector=nothing, order=Euclidean()) where {M} =
+  FilteredSearch{M}(method, nmax, maxpercategory, maxpersector, order)
+
+maxneighbors(method::FilteredSearch) = method.nmax
+
+function search!(neighbors, pₒ::Point, method::FilteredSearch; mask=nothing)
+  meth = method.method
+  obj  = meth.domain
+  inds = search(pₒ, meth, mask=mask)
+  nmax = method.nmax == 0 ? Inf : method.nmax
+
+  # get distances and give priority to closest neighbors if necessary
+  if !(meth isa KNearestSearch)
+    indcoords = reduce(hcat, coordinates.([centroid(obj, s) for s in inds]))
+    dists     = colwise(method.order, coordinates(pₒ), indcoords)
+    sorted    = sortperm(dists)
+    inds      = inds[sorted]
+  end
+
+  # initialize category and sectors constraints if necessary
+  categs  = initcategories(obj, inds, method.maxpercategory)
+  sectors = initsectors(meth, method.maxpersector)
+
+  # check about initialization size
+  nneigh = 0
+  for i in inds
+    nneigh == nmax && break
```

Comment:
Should this break be moved to the end of the loop?

Reply:
The final results will be the same, I can move there if you prefer

Reply:
I know, I think it is just a matter style :) The break of the while loop is usually placed near the end as we read the algorithm, specially because in this case we have `nneigh = 0` before the loop and the condition is clearly still false in the first iteration.
A break is usually placed at the beginning when you expect to save some computation.

Reply:
No problem, I'll change that. About the `nneigh = 0` and `nmax=0`, the "correct" default would be `nmax=Inf` in case only other constraints are desired. I set `nmax=0` and change it to `nmax = method.nmax == 0 ? Inf : method.nmax` in the beginning because the `search` method for all `BoundedNeighborSearchMethod` preallocate the `neighbors` size to `nmax` (https://github.com/JuliaGeometry/Meshes.jl/blob/master/src/neighborsearch.jl). Not a much elegant solution, but suggestions are welcome

Reply:
I need to find the time to review the code more carefully. More verbose comments could help here.